Realm/RLMSyncSubscription.mm

```diff
@@ -0,0 +1,523 @@
+    }
+
+    return [enumerator countByEnumeratingWithState:state count:len];
+}
+
+@implementation RLMSyncSubscriptionSet {
+    BOOL _isInWriteTransaction;
+    std::mutex _collectionEnumeratorMutex;
+    RLMRealm *_realm;
+}
+
+- (instancetype)initWithSubscriptionSet:(realm::sync::SubscriptionSet)subscriptionSet
+                                  realm:(RLMRealm *)realm {
+    if (self = [super init]) {
+        _subscriptionSet = std::make_unique<realm::sync::SubscriptionSet>(subscriptionSet);
+        _realm = realm;
+        return self;
+    }
+    return nil;
+}
+
+- (RLMSyncSubscriptionEnumerator *)fastEnumerator {
+    return [[RLMSyncSubscriptionEnumerator alloc] initWithSubscriptionSet:self];
+}
+
+- (void)unregisterEnumerator:(RLMSyncSubscriptionEnumerator *)enumerator {
+    std::lock_guard lock(_collectionEnumeratorMutex);
+    [_enumerators removeObject:enumerator];
+}
+
+- (void)registerEnumerator:(RLMSyncSubscriptionEnumerator *)enumerator {
+    std::lock_guard lock(_collectionEnumeratorMutex);
+    if (!_enumerators) {
+        _enumerators = [NSHashTable hashTableWithOptions:NSPointerFunctionsWeakMemory];
+    }
+    [_enumerators addObject:enumerator];
+}
+
+
+- (NSUInteger)count {
+    return _subscriptionSet->size();
+}
+
+- (nullable NSError *)error {
+    _subscriptionSet->refresh();
+    NSString *errorMessage = RLMStringDataToNSString(_subscriptionSet->error_str());
+    if (errorMessage.length == 0) {
+        return nil;
+    }
+    return [[NSError alloc] initWithDomain:RLMFlexibleSyncErrorDomain code:RLMFlexibleSyncErrorStatusError userInfo:@{NSLocalizedDescriptionKey: errorMessage}];
+}
+
+- (RLMSyncSubscriptionState)state {
+    _subscriptionSet->refresh();
+    switch (_subscriptionSet->state()) {
+        case realm::sync::SubscriptionSet::State::Uncommitted:
+        case realm::sync::SubscriptionSet::State::Pending:
+        case realm::sync::SubscriptionSet::State::Bootstrapping:
+            return RLMSyncSubscriptionStatePending;
+        case realm::sync::SubscriptionSet::State::Complete:
+            return RLMSyncSubscriptionStateComplete;
+        case realm::sync::SubscriptionSet::State::Error:
+            return RLMSyncSubscriptionStateError;
+        case realm::sync::SubscriptionSet::State::Superceded:
+            return RLMSyncSubscriptionStateSuperseded;
+    }
+}
+
+#pragma mark - Batch Update subscriptions
+
+- (void)write:(__attribute__((noescape)) void(^)(void))block {
+    return [self write:block onComplete:^(NSError*){}];
+}
+
+- (void)write:(__attribute__((noescape)) void(^)(void))block
+   onComplete:(void(^)(NSError *))completionBlock {
+    if (self->_isInWriteTransaction) {
+        @throw RLMException(@"Cannot initiate a write transaction on subscription set that is already been updated.");
+    }
+    _mutableSubscriptionSet = std::make_unique<realm::sync::MutableSubscriptionSet>(_subscriptionSet->make_mutable_copy());
+    self->_isInWriteTransaction = true;
```

Comment:
Having both `_mutableSubscriptionSet` and `_isInWriteTransaction` is redundant, and introduces the opportunity for bugs from them getting out of sync. We could instead check `_mutableSubscriptionSet` everywhere that we're currently checking `_isInWriteTransaction`.